### ISSUE TITLE
Pyright only complains about type Mock being compatible with other types if Mock is inside a predefined list

### ISSUE THREAD

**Issue by scorgn** [microsoft/pyright#2650]
### Describe the bug
I am using Pyright and it seems to not complain about Mock being incompatible with other types, except if a function requires a List of a specific type, and is passed a variable defined as `List[Mock]`. The same behavior is not seen if you're using `Tuple[...]` instead of `List[...]`, or even when directly passing in a list of Mock instances.

### Screenshots or Code
For example, in all of the following three snippets Pyright will not complain.
```python
from unittest.mock import Mock
 
 
class Foobar():
    pass


def some_function(foobar: Foobar):
    pass

mock_var: Mock = Mock()
some_function(mock_var)
```

```python
from unittest.mock import Mock
from typing import Tuple
 
 
class Foobar():
    pass


def some_function(foobar: Tuple[Foobar]):
    pass


tuple_var: Tuple[Mock] = (Mock(), )
some_function(tuple_var)
```

```python
from unittest.mock import Mock
from typing import List
 
 
class Foobar():
    pass


def some_function(foobar: List[Foobar]):
    pass


some_function([Mock()])
```

But in the following snippet, it will complain saying "Mock is incompatible with Foobar".

```python
from unittest.mock import Mock
from typing import List
 
 
class Foobar():
    pass


def some_function(foobar: List[Foobar]):
    pass


list_var: List[Mock] = [Mock()]
some_function(list_var)
```

Whether `list_var` is defined with the type declared (eg. `list_var: List[Mock] = [Mock()]`) or without the type declared (eg. `list_var = [Mock()]`) the result is the same.

This is the full error message:
```
[file]:14:15 - error: Argument of type "list[Mock]" cannot be assigned to parameter "foobar" of type "List[Foobar]" in function "some_function"
  TypeVar "_T@list" is invariant
    "Mock" is incompatible with "Foobar" (reportGeneralTypeIssues)
```

### Expected behavior
I would expect that passing in a predefined list of Mock instances would act the same way as both directly in a list of Mock instances and passing in a Tuple containing Mock instances.

### VS Code extension or command-line
This happens with both Pylance in VS Code, as well as Pyright in the CLI installed through npm. I've tested with both pyright v1.1.190 as well as pyright v1.1.191

### Additional context
[This seems](https://github.com/microsoft/pyright/issues/2618) to be partially related, but also a separate issue.

**Comment by erictraut** [comment]
The problem is that `Mock` derives from `NonCallableMock` which derives from `Any`. Any time there's an `Any` in a class's inheritance hierarchy, all bets are off in terms of type checking. 

The only place where a type checker can definitely emit an error is if the type is used as a type argument for an invariant type parameter, as with `List`. In this case, it's clearly an error to supply a `Mock` rather than `Foobar` because the types must match exactly. In your other examples above, `Mock` could be a subclass of the specified class, but the type checker can't determine this because `Any` appears in the class hierarchy.

Addressing this would require a change to the `mock.pyi` type stub in the typeshed repo. You might want to start a discussion there to see if such a change is feasible. I'm guessing that the `Any` base class was put there for a good reason and that removing it would result in a lot of false positive errors.

**Comment by scorgn** [comment]
Thanks for the response. I think I understand, so because there is no reference to `[Mock()]` it is safe to treat it as a covariant type - but when there is a reference to it, it becomes intravariant? I guess that makes sense, because if a `Foobar` object was added to the list from `some_function` it would no longer be of type `List[Mock]`, even if `Mock` is a subclass of `Foobar`.

> The problem is that Mock derives from NonCallableMock which derives from Any. Any time there's an Any in a class's inheritance hierarchy, all bets are off in terms of type checking.

To make sure I understand, the problem you're referencing is not that my fourth code snippet is throwing an error but that the first three code snippets *don't* throw an error?

I guess the actual solution to my issue is to define the function parameter type as `Sequence` or `Iterable` instead of `List` since I'm not actually altering the `List` in the function anyways, nor am I doing anything specific to Lists. Thanks again!
